src/app/ReadHandler.cpp

```diff
@@ -128,14 +170,30 @@ CHIP_ERROR ReadHandler::OnStatusReport(Messaging::ExchangeContext * apExchangeCo
 CHIP_ERROR ReadHandler::SendReportData(System::PacketBufferHandle && aPayload)
 {
     VerifyOrReturnLogError(IsReportable(), CHIP_ERROR_INCORRECT_STATE);
+    VerifyOrReturnLogError(mpExchangeCtx != nullptr, CHIP_ERROR_INCORRECT_STATE);
     if (IsInitialReport())
     {
         VerifyOrReturnLogError(mpExchangeCtx != nullptr, CHIP_ERROR_INCORRECT_STATE);
+        mSessionHandle.SetValue(mpExchangeCtx->GetSecureSession());
+    }
+    else
+    {
+        mpExchangeCtx = mpExchangeMgr->NewContext(mSessionHandle.Value(), this);
+        mpExchangeCtx->SetResponseTimeout(kImMessageTimeoutMsec);
     }
     VerifyOrReturnLogError(mpExchangeCtx != nullptr, CHIP_ERROR_INCORRECT_STATE);
-    MoveToState(HandlerState::Reporting);
-    return mpExchangeCtx->SendMessage(Protocols::InteractionModel::MsgType::ReportData, std::move(aPayload),
-                                      Messaging::SendFlags(Messaging::SendMessageFlags::kExpectResponse));
+    MoveToState(HandlerState::AwaitingReportResponse);
+    CHIP_ERROR err = mpExchangeCtx->SendMessage(Protocols::InteractionModel::MsgType::ReportData, std::move(aPayload),
```

Comment:
As discussed offline, when the sdk becomes spec compliant, it will be possible for the session to disappear out from under this exchange context in two scenarios: 1) session eviction and 2) session closing initiated by the peer.

This call chain will presumably fail in these cases, so we need to handle this if it is not handled already.  In this case, the subscription should be canceled.

Reply:
filed #9596,